runtime/test/recipe-resolver-test.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * @license
+ * Copyright (c) 2018 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import Arc from '../arc.js';
+import Loader from '../loader.js';
+import Manifest from '../manifest.js';
+import {RecipeResolver} from '../recipe/recipe-resolver.js';
+
+import {assert} from './chai-web.js';
+
+describe('RecipeResolver', function() {
+  const buildRecipe = async (content) => {
+    let registry = {};
+    let loader = new class extends Loader {
+      loadResource(path) {
+        return content[path];
+      }
+      path(fileName) {
+        return fileName;
+      }
+      join(_, file) {
+        return file;
+      }
+    };
+    let manifest = await Manifest.load('manifest', loader, {registry});
+    return manifest.recipes[0];
+  };
+
+  const createArc = () => {
+    return new Arc({
+      id: 'test',
+      slotComposer: {
+        affordance: 'dom',
+        getAvailableSlots: (() => {
+          return [{
+            name: 'root',
+            id: 'r0',
+            tags: ['#root'],
+            handles: [],
+            handleConnections: [],
+            getProvidedSlotSpec: () => {
+              return {isSet: false};
+            }
+          }];
+        })
+      }
+    });
+  };
+
+  it('resolves a recipe', async () => {
+    const arc = createArc();
+    const resolver = new RecipeResolver(arc);
+    const recipe = await buildRecipe({
+      manifest: `
+      particle P in 'A.js'
+        P()
+        consume root
+        affordance dom
+
+      recipe
+        P
+        `
+    });
+    // Initially the recipe should not be normalized (after which it's srozen).
+    assert.isFalse(Object.isFrozen(recipe));
+    const result = await resolver.resolve(recipe);
+    // The original recipe should remain untouched and the new instance
+    // should have been normalized.
+    assert.isFalse(Object.isFrozen(recipe));
+    assert.isTrue(Object.isFrozen(result));
+    assert.isTrue(result.isResolved());
+  });
+
+  it('returns null for an unresolvable recipe', async () => {
+    const arc = createArc();
+    const resolver = new RecipeResolver(arc);
+    const recipe = await buildRecipe({
+      manifest: `
+      particle P in 'A.js'
+        P(out * {Text value} text)
+
```

Comment:
Oddly if I add `consume root` and `affordance dom` to this, it seems to be returned as a resolved recipe even though there is no output handle. I've not looked at it further yet, thought welcome.

Reply:
Ah -- in that case it returns the recipe, but `isResolved` is false, and in this test case I was expecting `null`, updating shortly.